// @ant-design/charts/lib/graphs/radialTreeGraph/index.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var react_1 = __importStar(require("react"));
var g6_1 = __importDefault(require("@antv/g6"));
var base_1 = require("../../base");
var useGraph_1 = __importDefault(require("../../hooks/useGraph"));
var useProps_1 = __importDefault(require("../../hooks/useProps"));
var createLoading_1 = __importDefault(require("../../util/createLoading"));
var utils_1 = require("../utils");
var utils_2 = require("../utils");
var constants_1 = require("../constants");
var graphs = {};
var defaultLayout = {
    type: 'dendrogram',
    direction: 'LR',
    nodeSep: 20,
    rankSep: 100,
    radial: true,
};
var defaultProps = {
    nodeCfg: {
        type: 'circle',
        size: 30,
        anchorPoints: constants_1.defaultFlowGraphAnchorPoints,
        linkCenter: true,
        nodeStateStyles: constants_1.defaultStateStyles,
        style: constants_1.defaultNodeStyle,
    },
    edgeCfg: {
        type: 'line',
        edgeStateStyles: constants_1.defaultStateStyles,
    },
    behaviors: ['zoom-canvas', 'drag-canvas'],
    layout: defaultLayout,
    animate: true,
    markerPosition: 'right',
    autoFit: true,
    fitCenter: true,
    style: {
        height: 'inherit',
    },
};
var RadialTreeGraph = function (props) {
    var uProps = useProps_1.default(props, defaultProps).uProps;
    var data = uProps.data, className = uProps.className, style = uProps.style, width = uProps.width, height = uProps.height, behaviors = uProps.behaviors, layout = uProps.layout, animate = uProps.animate, nodeCfg = uProps.nodeCfg, edgeCfg = uProps.edgeCfg, autoFit = uProps.autoFit, fitCenter = uProps.fitCenter, minimapCfg = uProps.minimapCfg, onReady = uProps.onReady, loading = uProps.loading, loadingTemplate = uProps.loadingTemplate, errorTemplate = uProps.errorTemplate;
    var _a = nodeCfg !== null && nodeCfg !== void 0 ? nodeCfg : {}, nodeType = _a.type, nodeSize = _a.size, nodeAnchorPoints = _a.anchorPoints, nodeStateStyles = _a.nodeStateStyles, linkCenter = _a.linkCenter, nodeStyle = _a.style, nodeLabelCfg = _a.label;
    var _b = edgeCfg !== null && edgeCfg !== void 0 ? edgeCfg : {}, edgeType = _b.type, edgeStyle = _b.style, startArrowCfg = _b.startArrow, endArrowCfg = _b.endArrow, edgeLabelCfg = _b.label, edgeStateStyles = _b.edgeStateStyles;
    var container = react_1.default.useRef(null);
    var graph = react_1.default.useRef(null);
    var graphId = utils_1.getGraphId(graph);
    useGraph_1.default(graphs[graphId], uProps, container);
    react_1.useEffect(function () {
        var graphSize = utils_1.getGraphSize(width, height, container);
        var graph = graphs[graphId];
        if (!graph) {
            graph = new g6_1.default.TreeGraph({
                container: container.current,
                width: graphSize[0],
                height: graphSize[1],
                animate: animate,
                linkCenter: linkCenter,
                modes: {
                    default: behaviors,
                },
                defaultNode: {
                    type: nodeType,
                    size: nodeSize,
                    anchorPoints: nodeAnchorPoints,
                },
                defaultEdge: {
                    type: edgeType,
                },
                nodeStateStyles: nodeStateStyles,
                edgeStateStyles: edgeStateStyles,
                layout: layout,
                fitView: autoFit,
                fitCenter: fitCenter,
            });
            graphs[graphId] = graph;
        }
        graph.node(function (node) {
            var style = (nodeLabelCfg !== null && nodeLabelCfg !== void 0 ? nodeLabelCfg : {}).style;
            return {
                label: node.value,
                labelCfg: {
                    style: utils_2.getCommonConfig(style, node, graph),
                },
                style: __assign({ stroke: '#ccc' }, (typeof nodeStyle === 'function' ? nodeStyle(node, graph) : nodeStyle)),
            };
        });
        graph.edge(function (edge) {
            var startArrow = utils_2.getArrowCfg(startArrowCfg, edge);
            var endArrow = utils_2.getArrowCfg(endArrowCfg, edge);
            var _a = edgeLabelCfg !== null && edgeLabelCfg !== void 0 ? edgeLabelCfg : {}, style = _a.style, content = _a.content;
            return {
                label: utils_2.getCommonConfig(content, edge, graph),
                labelCfg: {
                    style: utils_2.getCommonConfig(style, edge, graph),
                },
                style: __assign({ stroke: '#ccc', startArrow: startArrow, endArrow: endArrow }, (typeof edgeStyle === 'function' ? edgeStyle(edge, graph) : edgeStyle)),
            };
        });
        utils_2.processMinimap(minimapCfg, graph);
        utils_2.bindStateEvents(graph, uProps);
        utils_2.renderGraph(graph, data);
        if (onReady) {
            onReady(graph);
        }
        return function () {
            if (graphs[graphId]) {
                graphs[graphId].destroy();
                delete graphs[graphId];
            }
        };
    }, []);
    return (react_1.default.createElement(base_1.ErrorBoundary, { errorTemplate: errorTemplate },
        loading && react_1.default.createElement(createLoading_1.default, { loadingTemplate: loadingTemplate }),
        react_1.default.createElement("div", { className: className, style: style, ref: container })));
};
exports.default = RadialTreeGraph;
