// @ant-design/charts/es/graphs/customItems.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
import G6 from '@antv/g6';
import { defaultMargin, defaultLabelStyle, defaultTitleLabelStyle, defaultTitleRectStyle, defaultCardStyle, defaultIconStyle, } from './constants';
import { getStyle, getCssPadding } from './utils';
import { createMarker } from './elements';
// 通用指标卡
export var registerIndicatorCardNode = function () {
    G6.registerNode('indicator-card', {
        // @ts-ignore
        draw: function (cfg, group) {
            if (cfg === void 0) { cfg = {}; }
            var _a = cfg.value, value = _a === void 0 ? {} : _a, nodeCfg = cfg.nodeCfg, markerCfg = cfg.markerCfg;
            var _b = nodeCfg, titleCfg = _b.title, itemsCfg = _b.items, _c = _b.label, label = _c === void 0 ? {} : _c, style = _b.style, _d = _b.padding, padding = _d === void 0 ? 0 : _d, customContent = _b.customContent;
            var labelStyle = label.style;
            var paddingArray = getCssPadding(padding);
            var _e = titleCfg !== null && titleCfg !== void 0 ? titleCfg : {}, titleStyle = _e.style, titleContainerStyle = _e.containerStyle;
            var _f = itemsCfg !== null && itemsCfg !== void 0 ? itemsCfg : {}, itemStyle = _f.style, itemContainerStyle = _f.containerStyle, layout = _f.layout, _g = _f.itemSpacing, itemSpacing = _g === void 0 ? 4 : _g, sort = _f.sort, _h = _f.padding, itemPadding = _h === void 0 ? [6, 0, 0] : _h;
            var itemPaddingArray = getCssPadding(itemPadding);
            var _j = value, title = _j.title, items = _j.items;
            var size = (cfg === null || cfg === void 0 ? void 0 : cfg.size) || [100, 30];
            if (typeof size === 'number')
                size = [size, size];
            var height = 0; // 统计容器总高度，动态设置
            var shapeWidth = size[0];
            var contentWidth = shapeWidth - paddingArray[1] - paddingArray[3];
            // card box
            var cardStyle = getStyle(style, cfg, group);
            var shape = group.addShape('rect', {
                attrs: __assign(__assign({ x: 0, y: 0, width: size[0], height: size[1] }, defaultCardStyle), cardStyle),
                name: 'main-box',
                draggable: true,
            });
            // node title
            var titleTextShape, itemShape, titleShape;
            if (title) {
                // title rect
                titleShape = group.addShape('rect', {
                    attrs: __assign(__assign({ x: 0, y: 0, width: size[0], height: 0 }, defaultTitleRectStyle), getStyle(titleContainerStyle, cfg, group)),
                    name: 'title-rect',
                    draggable: true,
                });
                titleTextShape = group.addShape('text', {
                    attrs: __assign(__assign({ x: paddingArray[3], y: paddingArray[0], textBaseline: 'top', text: title }, defaultTitleLabelStyle), getStyle(titleStyle, cfg, group)),
                    name: 'title',
                });
                var titleHeight = (titleTextShape
                    ? titleTextShape.getBBox()
                    : { height: size[1] / 2 }).height;
                titleShape === null || titleShape === void 0 ? void 0 : titleShape.attr('height', titleHeight + paddingArray[0] + paddingArray[2]);
                height += titleShape.getBBox().height;
            }
            if (items) {
                itemShape = group.addShape('rect', {
                    attrs: __assign({ x: paddingArray[3], y: height, width: contentWidth, height: 0 }, getStyle(itemContainerStyle, cfg, group)),
                    name: 'item-box',
                });
                height += itemPaddingArray[0];
                var itemContentWidth_1 = contentWidth - itemPaddingArray[1] - itemPaddingArray[3];
                var isArray_1 = Array.isArray(items);
                var createRowItems_1 = function (item, contentWidth, startX, index) {
                    if (index === void 0) { index = 0; }
                    var rowHeight = [];
                    var valueShapeWidth = 0;
                    var keys = sort ? Object.keys(item) : ['text', 'value', 'icon'];
                    keys.forEach(function (key, keyIndex) {
                        var x;
                        var isIcon = key.startsWith('icon');
                        // sort 直接均分，简单化
                        if (sort || layout === 'flex') {
                            x = (keyIndex * contentWidth) / keys.length;
                        }
                        else if (layout === 'follow') {
                            x = valueShapeWidth;
                        }
                        else {
                            // layout === 'bundled'
                            // 直接均分，icon 紧随 value
                            x = key === 'text' ? 0 : contentWidth / 2;
                            x += isIcon ? valueShapeWidth : 0;
                        }
                        var keyShape = group.addShape(isIcon ? 'image' : 'text', {
                            attrs: __assign(__assign({ textBaseline: 'top', x: startX + x, y: height, text: item[key], img: item[key] }, (isIcon ? defaultIconStyle : defaultLabelStyle)), getStyle(itemStyle || labelStyle, cfg, group, key)),
                            name: key + "-" + index + "-" + keyIndex,
                        });
                        if (key === 'value' || layout === 'follow') {
                            valueShapeWidth += keyShape.getBBox().width;
                            valueShapeWidth += layout === 'follow' ? itemSpacing : 0;
                        }
                        rowHeight.push(keyShape.getBBox().height);
                    });
                    return rowHeight;
                };
                var createItems_1 = function (item, index) {
                    var _a;
                    if (index === void 0) { index = 0; }
                    var itemsHeight = [];
                    if (customContent) {
                        itemsHeight.push((_a = customContent(item, group, {
                            startX: paddingArray[3] + itemPaddingArray[3],
                            startY: height,
                            width: itemContentWidth_1,
                        })) !== null && _a !== void 0 ? _a : 0);
                    }
                    else {
                        itemsHeight.push.apply(itemsHeight, createRowItems_1(item, itemContentWidth_1, paddingArray[3] + itemPaddingArray[3], index));
                    }
                    height += Math.max.apply(Math, itemsHeight);
                    if (isArray_1 && index !== items.length - 1) {
                        height += defaultMargin;
                    }
                };
                if (Array.isArray(items)) {
                    items.forEach(function (item, index) {
                        createItems_1(item, index);
                    });
                }
                else {
                    createItems_1(items);
                }
            }
            itemShape === null || itemShape === void 0 ? void 0 : itemShape.attr('height', Math.max(height - (titleShape === null || titleShape === void 0 ? void 0 : titleShape.getBBox().height) + itemPaddingArray[2], size[1]));
            shape === null || shape === void 0 ? void 0 : shape.attr('height', items
                ? (titleShape === null || titleShape === void 0 ? void 0 : titleShape.getBBox().height) + (itemShape === null || itemShape === void 0 ? void 0 : itemShape.getBBox().height) + paddingArray[2]
                : (titleShape === null || titleShape === void 0 ? void 0 : titleShape.getBBox().height) + (itemShape === null || itemShape === void 0 ? void 0 : itemShape.getBBox().height));
            // collapse marker
            if (markerCfg) {
                var _k = shape.getBBox(), shapeWidth_1 = _k.width, shapeHeight = _k.height;
                var _l = typeof markerCfg === 'function' ? markerCfg(cfg, group) : markerCfg, show = _l.show, _m = _l.position, position = _m === void 0 ? 'right' : _m, collapsed = _l.collapsed, markerStyle = _l.style;
                createMarker({
                    show: show,
                    position: position,
                    collapsed: collapsed,
                    style: markerStyle,
                }, group, [shapeWidth_1, shapeHeight]);
                shape.attr('defaultCollapsed', collapsed);
            }
            return shape;
        },
        /**
         * 更新节点，包含文本
         * @override
         * @param  {Object} cfg 节点的配置项
         * @param  {Node} node 节点
         */
        update: function (cfg, node) {
            var group = node.getContainer();
            var markerShape = group
                .get('children')
                .find(function (item) { return item.get('type') === 'marker'; });
            var collapsed = node.getModel().collapsed;
            markerShape === null || markerShape === void 0 ? void 0 : markerShape.attr({
                symbol: collapsed ? G6.Marker.expand : G6.Marker.collapse,
            });
        },
    }, 'single-node');
};
// 组织架构图
export var registerOrganizationCardNode = function () {
    G6.registerNode('organization-card', {
        draw: function (cfg, group) {
            if (cfg === void 0) { cfg = {}; }
            var _a = cfg.value, value = _a === void 0 ? {} : _a, nodeCfg = cfg.nodeCfg, markerCfg = cfg.markerCfg;
            var _b = nodeCfg, style = _b.style, _c = _b.padding, padding = _c === void 0 ? 0 : _c, _d = _b.label, label = _d === void 0 ? {} : _d, customContent = _b.customContent;
            var labelStyle = label.style;
            var paddingArray = getCssPadding(padding);
            var size = ((cfg === null || cfg === void 0 ? void 0 : cfg.size) || [100, 30]);
            if (typeof size === 'number')
                size = [size, size];
            var height = 0; // 统计容器总高度，动态设置，宽度不做调整
            var contentWidth = size[0] - paddingArray[1] - paddingArray[3];
            // card box
            var cardStyle = getStyle(style, cfg, group);
            var shape = group.addShape('rect', {
                attrs: __assign(__assign({ x: 0, y: 0, width: size[0], height: size[1] }, defaultCardStyle), cardStyle),
                name: 'main-box',
                draggable: true,
            });
            if (value) {
                height += paddingArray[0];
                var createRowItems_2 = function (item, contentWidth, startX, index) {
                    if (index === void 0) { index = 0; }
                    var iconWidth = 0;
                    var rowHeight = [];
                    var keys = ['icon', 'text', 'value'];
                    var getXY = function (type, layoutCfg) {
                        var _a = layoutCfg.fontSize, fontSize = _a === void 0 ? 12 : _a;
                        var x = 0;
                        var y = 0;
                        switch (type) {
                            case 'icon':
                                x = startX;
                                y = height;
                                break;
                            case 'text':
                                x = startX + (contentWidth + iconWidth) / 2;
                                y = item.value ? paddingArray[0] : (size[1] - fontSize) / 2;
                                break;
                            case 'value':
                                x = startX + (contentWidth + iconWidth) / 2;
                                y = item.text
                                    ? paddingArray[0] + rowHeight[1] + defaultMargin
                                    : (size[1] - fontSize) / 2;
                                break;
                            default:
                                break;
                        }
                        return { x: x, y: y };
                    };
                    keys.forEach(function (key, keyIndex) {
                        var isIcon = key.startsWith('icon');
                        var shapeStyle = getStyle(labelStyle, cfg, group, key);
                        if (key === 'icon' && item[key]) {
                            iconWidth = shapeStyle.width || 32;
                        }
                        var keyShape = group.addShape(isIcon ? 'image' : 'text', {
                            attrs: __assign(__assign(__assign(__assign({ textBaseline: 'top', textAlign: 'center' }, getXY(key, shapeStyle)), { text: item[key], img: item[key] }), (isIcon ? defaultIconStyle : defaultLabelStyle)), shapeStyle),
                            name: key + "-" + index + "-" + keyIndex,
                        });
                        rowHeight.push(keyShape.getBBox().height);
                    });
                    return rowHeight;
                };
                var createItems = function (item, index) {
                    var _a;
                    if (index === void 0) { index = 0; }
                    var itemsHeight = [];
                    if (customContent) {
                        itemsHeight.push((_a = customContent(item, group, {
                            startX: paddingArray[3],
                            startY: height,
                            width: contentWidth,
                        })) !== null && _a !== void 0 ? _a : 0);
                    }
                    else {
                        itemsHeight.push.apply(itemsHeight, createRowItems_2(item, contentWidth, paddingArray[3], index));
                    }
                    height += Math.max.apply(Math, itemsHeight);
                };
                createItems(value);
            }
            shape === null || shape === void 0 ? void 0 : shape.attr('height', Math.max(height + paddingArray[2], size[1]));
            // collapse marker
            if (markerCfg) {
                var _e = shape.getBBox(), shapeWidth = _e.width, shapeHeight = _e.height;
                var _f = typeof markerCfg === 'function' ? markerCfg(cfg, group) : markerCfg, show = _f.show, _g = _f.position, position = _g === void 0 ? 'right' : _g, collapsed = _f.collapsed, markerStyle = _f.style;
                createMarker({
                    show: show,
                    position: position,
                    collapsed: collapsed,
                    style: markerStyle,
                }, group, [shapeWidth, shapeHeight]);
                shape.attr('defaultCollapsed', collapsed);
            }
            return shape;
        },
    }, 'single-node');
};
