// @ant-design/charts/es/graphs/decompositionTreeGraph/index.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
import React, { useEffect } from 'react';
import G6 from '@antv/g6';
import { ErrorBoundary } from '../../base';
import useGraph from '../../hooks/useGraph';
import useProps from '../../hooks/useProps';
import ChartLoading from '../../util/createLoading';
import { getGraphSize, getGraphId, processMinimap, renderGraph, getCommonConfig, getArrowCfg, bindStateEvents, bindDefaultEvents, } from '../utils';
import { defaultFlowGraphAnchorPoints, defaultNodeSize, defaultStateStyles, defaultNodeStyle, } from '../constants';
import { registerIndicatorCardNode } from '../flowAnalysisGraph/customItem';
var graphs = {};
registerIndicatorCardNode();
var defaultLayout = {
    type: 'compactBox',
    direction: 'LR',
    getId: function (d) {
        return d.id;
    },
    getHeight: function () {
        return 60;
    },
    getWidth: function () {
        return 16;
    },
    getVGap: function () {
        return 16;
    },
    getHGap: function () {
        return 100;
    },
};
var defaultProps = {
    nodeCfg: {
        type: 'indicator-card',
        size: defaultNodeSize,
        style: defaultNodeStyle,
        anchorPoints: defaultFlowGraphAnchorPoints,
        padding: 6,
        layout: 'bundled',
        nodeStateStyles: defaultStateStyles,
        label: {
            style: function (cfg, group, type) {
                var styles = {
                    icon: {
                        width: 10,
                        height: 10,
                    },
                    value: {
                        fill: '#000',
                    },
                    text: {
                        fill: '#aaa',
                    },
                };
                return type ? styles[type] : {};
            },
        },
    },
    edgeCfg: {
        type: 'cubic-horizontal',
        endArrow: {
            type: 'vee',
        },
        edgeStateStyles: defaultStateStyles,
    },
    behaviors: ['zoom-canvas', 'drag-canvas'],
    layout: defaultLayout,
    animate: true,
    autoFit: true,
    fitCenter: true,
    style: {
        height: 'inherit',
    },
};
var DecompositionTreeGraph = function (props) {
    var uProps = useProps(props, defaultProps).uProps;
    var data = uProps.data, className = uProps.className, style = uProps.style, width = uProps.width, height = uProps.height, _a = uProps.behaviors, behaviors = _a === void 0 ? ['zoom-canvas', 'drag-canvas'] : _a, layout = uProps.layout, animate = uProps.animate, edgeCfg = uProps.edgeCfg, nodeCfg = uProps.nodeCfg, markerCfg = uProps.markerCfg, minimapCfg = uProps.minimapCfg, autoFit = uProps.autoFit, fitCenter = uProps.fitCenter, onReady = uProps.onReady, loading = uProps.loading, loadingTemplate = uProps.loadingTemplate, errorTemplate = uProps.errorTemplate;
    var _b = nodeCfg !== null && nodeCfg !== void 0 ? nodeCfg : {}, nodeType = _b.type, nodeSize = _b.size, nodeAnchorPoints = _b.anchorPoints, nodeStateStyles = _b.nodeStateStyles, nodeStyle = _b.style, nodeLabelCfg = _b.title;
    var _c = edgeCfg !== null && edgeCfg !== void 0 ? edgeCfg : {}, edgeType = _c.type, edgeStyle = _c.style, startArrowCfg = _c.startArrow, endArrowCfg = _c.endArrow, labelCfg = _c.label, edgeStateStyles = _c.edgeStateStyles;
    var container = React.useRef(null);
    var graph = React.useRef(null);
    var graphId = getGraphId(graph);
    useGraph(graphs[graphId], uProps, container);
    useEffect(function () {
        var graphSize = getGraphSize(width, height, container);
        var graph = graphs[graphId];
        if (!graph) {
            graph = new G6.TreeGraph({
                container: container.current,
                width: graphSize[0],
                height: graphSize[1],
                animate: animate,
                modes: {
                    default: behaviors,
                },
                defaultNode: {
                    type: nodeType,
                    size: nodeSize,
                    anchorPoints: nodeAnchorPoints,
                    nodeCfg: nodeCfg,
                },
                defaultEdge: {
                    type: edgeType,
                    edgeCfg: edgeCfg,
                },
                nodeStateStyles: nodeStateStyles,
                edgeStateStyles: edgeStateStyles,
                layout: layout,
                fitView: autoFit,
                fitCenter: fitCenter,
            });
            graphs[graphId] = graph;
        }
        // defaultNode 默认只能绑定 plainObject，针对 Function 类型需要通过该模式绑定
        graph.node(function (node) {
            var _a;
            if (nodeType === 'indicator-card') {
                node.markerCfg = markerCfg;
                return {};
            }
            var style = nodeLabelCfg.style;
            return {
                label: (_a = node.value) === null || _a === void 0 ? void 0 : _a.title,
                labelCfg: {
                    style: getCommonConfig(style, node, graph),
                },
                style: __assign({}, (typeof nodeStyle === 'function' ? nodeStyle(node, graph) : nodeStyle)),
            };
        });
        graph.edge(function (edge) {
            var startArrow = getArrowCfg(startArrowCfg, edge);
            var endArrow = getArrowCfg(endArrowCfg, edge);
            var style = (labelCfg !== null && labelCfg !== void 0 ? labelCfg : {}).style;
            return {
                label: edge.value,
                labelCfg: {
                    style: getCommonConfig(style, edge, graph),
                },
                style: __assign({ stroke: '#ccc', startArrow: startArrow, endArrow: endArrow }, (typeof edgeStyle === 'function' ? edgeStyle(edge, graph) : edgeStyle)),
            };
        });
        processMinimap(minimapCfg, graph);
        bindStateEvents(graph, uProps);
        if (markerCfg) {
            bindDefaultEvents(graph);
        }
        renderGraph(graph, data);
        if (onReady) {
            onReady(graph);
        }
        return function () {
            if (graphs[graphId]) {
                graphs[graphId].destroy();
                delete graphs[graphId];
            }
        };
    }, []);
    return (React.createElement(ErrorBoundary, { errorTemplate: errorTemplate },
        loading && React.createElement(ChartLoading, { loadingTemplate: loadingTemplate }),
        React.createElement("div", { className: className, style: style, ref: container })));
};
export default DecompositionTreeGraph;
