// @ant-design/charts/lib/graphs/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.defaultIconStyle = exports.defaultTitleRectStyle = exports.defaultTitleLabelStyle = exports.defaultMargin = exports.defaultCardStyle = exports.defaultMinimapCfg = exports.defaultLabelStyle = exports.defaultFlowGraphAnchorPoints = exports.defaultNodeStyle = exports.defaultNodeSize = exports.defaultStateStyles = void 0;
// 默认交互状态
exports.defaultStateStyles = {
    hover: {
        stroke: '#1890ff',
        lineWidth: 2,
    },
};
// card 默认节点大小
exports.defaultNodeSize = [120, 40];
// 默认节点样式
exports.defaultNodeStyle = {
    stroke: '#40a9ff',
};
// 默认 anchor 连接点
exports.defaultFlowGraphAnchorPoints = [
    [0, 0.5],
    [1, 0.5],
];
// card body|footer  默认样式
exports.defaultLabelStyle = {
    fill: '#000',
    fontSize: 12,
};
exports.defaultMinimapCfg = {
    show: false,
    size: [150, 100],
    type: 'keyShape',
};
// card title 默认样式
exports.defaultCardStyle = {
    fill: '#fff',
    stroke: '#40a9ff',
    radius: 2,
};
// card 内部 padding | margin | 行间距
exports.defaultMargin = 6;
// title 默认样式
exports.defaultTitleLabelStyle = {
    fill: '#fff',
    fontSize: 12,
};
// title 容器默认样式
exports.defaultTitleRectStyle = {
    fill: '#40a9ff',
    radius: [2, 2, 0, 0],
};
// card icon 默认样式
exports.defaultIconStyle = {
    width: 12,
    height: 12,
};
