// @ant-design/charts/lib/graphs/decompositionTreeGraph/index.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var react_1 = __importStar(require("react"));
var g6_1 = __importDefault(require("@antv/g6"));
var base_1 = require("../../base");
var useGraph_1 = __importDefault(require("../../hooks/useGraph"));
var useProps_1 = __importDefault(require("../../hooks/useProps"));
var createLoading_1 = __importDefault(require("../../util/createLoading"));
var utils_1 = require("../utils");
var constants_1 = require("../constants");
var customItem_1 = require("../flowAnalysisGraph/customItem");
var graphs = {};
customItem_1.registerIndicatorCardNode();
var defaultLayout = {
    type: 'compactBox',
    direction: 'LR',
    getId: function (d) {
        return d.id;
    },
    getHeight: function () {
        return 60;
    },
    getWidth: function () {
        return 16;
    },
    getVGap: function () {
        return 16;
    },
    getHGap: function () {
        return 100;
    },
};
var defaultProps = {
    nodeCfg: {
        type: 'indicator-card',
        size: constants_1.defaultNodeSize,
        style: constants_1.defaultNodeStyle,
        anchorPoints: constants_1.defaultFlowGraphAnchorPoints,
        padding: 6,
        layout: 'bundled',
        nodeStateStyles: constants_1.defaultStateStyles,
        label: {
            style: function (cfg, group, type) {
                var styles = {
                    icon: {
                        width: 10,
                        height: 10,
                    },
                    value: {
                        fill: '#000',
                    },
                    text: {
                        fill: '#aaa',
                    },
                };
                return type ? styles[type] : {};
            },
        },
    },
    edgeCfg: {
        type: 'cubic-horizontal',
        endArrow: {
            type: 'vee',
        },
        edgeStateStyles: constants_1.defaultStateStyles,
    },
    behaviors: ['zoom-canvas', 'drag-canvas'],
    layout: defaultLayout,
    animate: true,
    autoFit: true,
    fitCenter: true,
    style: {
        height: 'inherit',
    },
};
var DecompositionTreeGraph = function (props) {
    var uProps = useProps_1.default(props, defaultProps).uProps;
    var data = uProps.data, className = uProps.className, style = uProps.style, width = uProps.width, height = uProps.height, _a = uProps.behaviors, behaviors = _a === void 0 ? ['zoom-canvas', 'drag-canvas'] : _a, layout = uProps.layout, animate = uProps.animate, edgeCfg = uProps.edgeCfg, nodeCfg = uProps.nodeCfg, markerCfg = uProps.markerCfg, minimapCfg = uProps.minimapCfg, autoFit = uProps.autoFit, fitCenter = uProps.fitCenter, onReady = uProps.onReady, loading = uProps.loading, loadingTemplate = uProps.loadingTemplate, errorTemplate = uProps.errorTemplate;
    var _b = nodeCfg !== null && nodeCfg !== void 0 ? nodeCfg : {}, nodeType = _b.type, nodeSize = _b.size, nodeAnchorPoints = _b.anchorPoints, nodeStateStyles = _b.nodeStateStyles, nodeStyle = _b.style, nodeLabelCfg = _b.title;
    var _c = edgeCfg !== null && edgeCfg !== void 0 ? edgeCfg : {}, edgeType = _c.type, edgeStyle = _c.style, startArrowCfg = _c.startArrow, endArrowCfg = _c.endArrow, labelCfg = _c.label, edgeStateStyles = _c.edgeStateStyles;
    var container = react_1.default.useRef(null);
    var graph = react_1.default.useRef(null);
    var graphId = utils_1.getGraphId(graph);
    useGraph_1.default(graphs[graphId], uProps, container);
    react_1.useEffect(function () {
        var graphSize = utils_1.getGraphSize(width, height, container);
        var graph = graphs[graphId];
        if (!graph) {
            graph = new g6_1.default.TreeGraph({
                container: container.current,
                width: graphSize[0],
                height: graphSize[1],
                animate: animate,
                modes: {
                    default: behaviors,
                },
                defaultNode: {
                    type: nodeType,
                    size: nodeSize,
                    anchorPoints: nodeAnchorPoints,
                    nodeCfg: nodeCfg,
                },
                defaultEdge: {
                    type: edgeType,
                    edgeCfg: edgeCfg,
                },
                nodeStateStyles: nodeStateStyles,
                edgeStateStyles: edgeStateStyles,
                layout: layout,
                fitView: autoFit,
                fitCenter: fitCenter,
            });
            graphs[graphId] = graph;
        }
        // defaultNode 默认只能绑定 plainObject，针对 Function 类型需要通过该模式绑定
        graph.node(function (node) {
            var _a;
            if (nodeType === 'indicator-card') {
                node.markerCfg = markerCfg;
                return {};
            }
            var style = nodeLabelCfg.style;
            return {
                label: (_a = node.value) === null || _a === void 0 ? void 0 : _a.title,
                labelCfg: {
                    style: utils_1.getCommonConfig(style, node, graph),
                },
                style: __assign({}, (typeof nodeStyle === 'function' ? nodeStyle(node, graph) : nodeStyle)),
            };
        });
        graph.edge(function (edge) {
            var startArrow = utils_1.getArrowCfg(startArrowCfg, edge);
            var endArrow = utils_1.getArrowCfg(endArrowCfg, edge);
            var style = (labelCfg !== null && labelCfg !== void 0 ? labelCfg : {}).style;
            return {
                label: edge.value,
                labelCfg: {
                    style: utils_1.getCommonConfig(style, edge, graph),
                },
                style: __assign({ stroke: '#ccc', startArrow: startArrow, endArrow: endArrow }, (typeof edgeStyle === 'function' ? edgeStyle(edge, graph) : edgeStyle)),
            };
        });
        utils_1.processMinimap(minimapCfg, graph);
        utils_1.bindStateEvents(graph, uProps);
        if (markerCfg) {
            utils_1.bindDefaultEvents(graph);
        }
        utils_1.renderGraph(graph, data);
        if (onReady) {
            onReady(graph);
        }
        return function () {
            if (graphs[graphId]) {
                graphs[graphId].destroy();
                delete graphs[graphId];
            }
        };
    }, []);
    return (react_1.default.createElement(base_1.ErrorBoundary, { errorTemplate: errorTemplate },
        loading && react_1.default.createElement(createLoading_1.default, { loadingTemplate: loadingTemplate }),
        react_1.default.createElement("div", { className: className, style: style, ref: container })));
};
exports.default = DecompositionTreeGraph;
